src/gflownet/data/sampling_iterator.py

```diff
@@ -201,49 +218,97 @@ def __iter__(self):
                     if self.log_molecule_smis:
                         for i, m in zip(valid_idcs, valid_mols):
                             trajs[i]["smi"] = Chem.MolToSmiles(m)
-            flat_rewards = torch.stack(flat_rewards)
+
             # Compute scalar rewards from conditional information & flat rewards
+            flat_rewards = torch.stack(flat_rewards)
             log_rewards = self.task.cond_info_to_logreward(cond_info, flat_rewards)
             log_rewards[torch.logical_not(is_valid)] = self.algo.illegal_action_logreward
-            # Construct batch
-            batch = self.algo.construct_batch(trajs, cond_info["encoding"], log_rewards)
-            batch.num_offline = num_offline
-            batch.num_online = num_online
-            batch.flat_rewards = flat_rewards
-            batch.mols = mols
-            batch.preferences = cond_info.get("preferences", None)
-            # TODO: we could very well just pass the cond_info dict to construct_batch above,
-            # and the algo can decide what it wants to put in the batch object
 
+            # Computes some metrics
             if not self.sample_cond_info:
                 # If we're using a dataset of preferences, the user may want to know the id of the preference
                 for i, j in zip(trajs, idcs):
                     i["data_idx"] = j
-
-            # Converts back into natural rewards for logging purposes
-            # (allows to take averages and plot in objective space)
-            # TODO: implement that per-task (in case they don't apply the same beta and log transformations)
+            #  note: we convert back into natural rewards for logging purposes
+            #  (allows to take averages and plot in objective space)
+            #  TODO: implement that per-task (in case they don't apply the same beta and log transformations)
             rewards = torch.exp(log_rewards / cond_info["beta"])
-
             if num_online > 0 and self.log_dir is not None:
                 self.log_generated(
-                    trajs[num_offline:],
-                    rewards[num_offline:],
-                    flat_rewards[num_offline:],
-                    {k: v[num_offline:] for k, v in cond_info.items()},
+                    deepcopy(trajs[num_offline:]),
```

Comment:
Were those to temporarily try guarding against replay buffer shenanigans? Or do things break if we remove the `deepcopy`s?

Reply:
I had a different (but most probably related) issue with monitoring; the focus_accuracy metric was wrong probably because `flat_rewards` was getting modified somehow, I didn't take the time to thoroughly investigate but there is an issue opened [here](https://github.com/recursionpharma/gflownet/issues/61) and my guess is that solving it would allow us to remove deepcopy's eveywhere in sampling_iterator (both for the replay buffer and the metrics).

I also documented this [here](https://docs.google.com/presentation/d/1VpnorJ1SIJQR54vO1ZnOoKpfGY8ysF_lvl0zQmPIhxo/edit?usp=sharing), see slide on April 10th.